### ISSUE TITLE
vertx eventloop is slowing consuming from eventbus on load

### ISSUE THREAD

**Issue by IdanFridman** [eclipse-vertx/vert.x#2779]
We are polling messages from kafka (using Executor Thread) and put it on vert.x eventbus. eventually the verticles(non-workers) consuming those messages from the eventbus.

we measured that the time it takes to each message to be consumed from vertx eventbus after it was sent to it can reach to 4 seconds on peaks.

for some reason occasionally it takes to the eventloop's verticle consumer long time to consume messages still but not showing the famous thread block error anywhere else. So we cant be sure what delaying it(we set the threadblock warning to be on 100ms)

weird thing that we dont see any event loop block warnings. cpu/ram are all good.

we measured the time it takes for a messages from the time we send it to the eventbus until it was subscribed and it took around 1 second. on average our event-loop job metrics shows around 60ms so no way eventloop is being delayed picking up messages.

perhaps something regarding the eventbus ?

thanks

**Comment by IdanFridman** [comment]
@vietj iam using our own kafka client. we have one executor thread that takes from kafka and put it straight away on eventbus. then we have on other side an eventloop that should take it and process it

**Comment by IdanFridman** [comment]
have you monitored GC pauses. no. can we have instructions what exactly we should monitor on that subject or refer me to some link/known issue ?

**Comment by IdanFridman** [comment]
@vietj I also want to mention that the load is 40 ms/s thats pretty much low load (this latency happened to us after many hours.. not always)

**Comment by vietj** [comment]
I'm assuming this is local ?

one thing you can do is periodically (every 1s), do in each verticle

```
long now = System.currentMillis();
context.runOnContext(v -> {
   long elapsed = System.currentMillis() - now;
});
```

This will measure in each verticle, how long it takes to execute a task on the verticle event loop.

This number will give us an idea of each event loop latency.

**Comment by IdanFridman** [comment]
@vietj in the beginning with used localConsumer (if thats what you refer when you ask "this is local..") then we thought this might do some performances issues so we switched back to regular event bus consumer which didnt demonstrate any difference

we will add this to one of our services 
long now = System.currentMillis();
context.runOnContext(v -> {
   long elapsed = System.currentMillis() - now;
});

to our metrics and come back with results. if you think about anything else we shall measure check please note

**Comment by vietj** [comment]
@IdanFridman you can have your event loop saturated with tasks, can you check whether the spike you see is related to network burst ?

if your event loop is  getting saturated, what you need to do is deploy several instances of you verticle so it's scaled to more than one event loop.

**Comment by IdanFridman** [comment]
@vietj we tried something else. we made our eventbus within non clustered mode. seems like pickink-up fron the eventbus became must faster. I afraid eventbus within cluster-mode doesnt perform as expected. we are talking about 40ms/s which consider as  relatively low

**Comment by vietj** [comment]
yes this is low and there must be something unexpected happening. do you have some kind of ack mechanism regarding reading kafka messages that might slow this down ?

**Comment by vietj** [comment]
do you have a reproducer by the way ? I would be glad having a look at it.

**Comment by IdanFridman** [comment]
@vietj the "ack" mechanism with kafka happening before we hand-off the messages to the eventbus 
meaning it happens on the the executor thread scope so it doesn't delay or interrupted the eventbus/eventloop in anyway

regarding re-producer it will be hard to do it as we run it for 8 hours and we see the peaks times to times. but i can pass some code so you understand the flow:

https://gist.github.com/IdanFridman/1aa29e3126e07d0049836f2ee380e19a

thanks.

**Comment by IdanFridman** [comment]
@vietj I was wondering if you had a chance to look at my code example ? we really frustrated over here with the eventloop performances

**Comment by vietj** [comment]
@IdanFridman looking at the code I see nothing that would create such latencies besides GC pauses, so I think you should really investigate those and also the liveliness of the event-loop threads (i.e are these latency spikes due to Vertx or due to the fact that the event-loop threads are suspended for some reason (like full GC)).

**Comment by IdanFridman** [comment]
@vietj we havnt seen anything special regarding GC's also we shall have seen event-loop blocked warning in such case no? we havnt got any of those (our eventloop loop warnings tuned to raise on block of 50ms)

throwing this on GC's stop is kind of shooting in the air with a hope to catch something..

perhaps we can monitor something else.. perhaps the HTTP requests being busy.. but they also didnt throw any warnings

**Comment by vietj** [comment]
I think you should look for anything that has a burst effect in your application.

that being said, the event-loop is fair, so to have a starvation of the event loop for the event bus, means that many event loop tasks gets created simultaneously.

you could try modify the vertx source code and keep in a log the tasks.toString() along with the current time.

when you experience a latency spike you look at the tasks that were executed in that time range to have an idea of what is going on.

**Comment by IdanFridman** [comment]
@vietj modified now vertx source code seems like over-rated for what we want to achieve. I want to demonstrate current situation where we now working on non-cluster-mode for the eventbus and we have only workers that take messages from the eventbus. we measured the worker time from start to end and it's very very low. however many messages in eventbus being queued-up.

how messages are being passed from the eventbus to the worker verticle? is it by an executor thread or the eventloop it self?
thanks.

**Comment by ashishlamba** [comment]
We are also facing same issue, instead of event loop we are using worker verticle for the processing. Any luck on this?

**Comment by ashishlamba** [comment]
> @vietj modified now vertx source code seems like over-rated for what we want to achieve. I want to demonstrate current situation where we now working on non-cluster-mode for the eventbus and we have only workers that take messages from the eventbus. we measured the worker time from start to end and it's very very low. however many messages in eventbus being queued-up.
> 
> how messages are being passed from the eventbus to the worker verticle? is it by an executor thread or the eventloop it sel
> thanks.



> @vietj modified now vertx source code seems like over-rated for what we want to achieve. I want to demonstrate current situation where we now working on non-cluster-mode for the eventbus and we have only workers that take messages from the eventbus. we measured the worker time from start to end and it's very very low. however many messages in eventbus being queued-up.
> 
> how messages are being passed from the eventbus to the worker verticle? is it by an executor thread or the eventloop it self?
> thanks.

We are also facing same issue, instead of event loop we are using worker verticle for the processing. Any luck on this?

**Comment by vietj** [comment]
can you provide a reproducer @ashishlamba ?

**Comment by ashishlamba** [comment]
As per our requirement we need to call a blocking API which takes ~400-500ms. For this  we are using worker verticle so that event loop is not blocked. We are putting messages in event bus from the main verticle and consuming in the worker verticle. Currently we are running 20 instances of the worker verticle and load is 20-40 tps only still we are getting event bus message reply timeouts on the main verticle. For 3 lakh events we got 0-400 timeouts for different runs. (btw default event bus message reply timeout is 30sec and we haven't changed it). It looks like messages are not processing as fast as the load tps and some taking 20-40 secs from event bus send to consume, hence resulting in timeouts.

**Comment by ashishlamba** [comment]
As per our requirement we need to call a blocking API which have ~400-500ms response time. For this we are using worker verticle so that event loop is not blocked. We are putting messages in event bus from the main verticle and consuming in the worker verticle. Currently we are running 20 instances of the worker verticle and application load is 20-40 tps only still we are getting event bus message reply timeouts on the main verticle. For 3 lakh events we got 0-400 timeouts for different runs. (btw default event bus message reply timeout is 30sec and we haven't changed it). It looks like messages are not processing as fast as the load tps and some taking 20-40 secs from main verticle event bus send to the worker verticle consumer handler, hence resulting in timeouts.

**Comment by vietj** [comment]
can you provide a reproducer for this ?

I would also advocate to not use worker verticles with EventBus and instead use regular verticle. In this verticle you can use executeBlocking that should be easier to handle load, i.e try something like:

```java
eventBus.consumer("my.address", msg -> {
  vertx.executeBlocking(promise -> {
    // Do something blocking
    promise.complete(result);
  }, ar -> {
    if (ar.succeeded()) {
      msg.reply("it worked");
    } else {
      msg.fail("it failed");
    }
});
});
```

**Comment by asad-awadia** [comment]
@ashishlamba can you share some code for that test? So for every 300k events, you are getting up to 400 errors/failures?

**Comment by Aman-Garg** [comment]
@asad-awadia @ashishlamba @vietj @IdanFridman @tobias 
any solution on this, I am also facing exact same issue in my application

**Comment by asad-awadia** [comment]
@Aman-Garg post the code that writes to the eventbus and the part that reads from it - can't help without seeing what you are doing

**Comment by Aman-Garg** [comment]
Thanks @asad-awadia
As per organization guidelines I cannot paste the exact code here 
But to send an event to the bus we are using - eventBus.rxRequest(Address_A, Message, DeliveryOptions);
Reading it like - vertx.eventBus().consumer(Address_A, this::method());

**Comment by alessandrasonsini** [comment]
> the ELPickupTime is another metric which shows the latency how long it took for a message to be consumed from the eventbus by the eventlopp

Can i ask you how did you create this metric? i'm facing a similan issue and to analyze the latency could really help me! thanks
